### ISSUE TITLE
Issue with top_file_merging_strategy same

### ISSUE THREAD

**Issue by ccoughle** [saltstack/salt#43293]
### Description of Issue/Question

We have about 10 different environments set for the machines we manage.  File roots look like this:

```
file_roots:
  base:
    - /srv/salt/base
  dev:
    - /srv/salt/dev
    - /srv/salt/base
    - /srv/salt/common
  env1:
    - /srv/salt/env1
    - /srv/salt/base
    - /srv/salt/common
```

our merge strategy is set to same.  Environments are configured on the minions in minion.d
This works well for us, allow shared states between environments along with isolation of environments.  However, we had an issue where the environments specified on the minions got blanked out.  So, in absence of a defined environment; all the top files in all environments were merged.  

The "base" environment's top file contains only a single state that manages the salt minion's configs and applies configs, such as the minion's defined environment, based on an external pillar source (cobbler in this case).

Our expectation was that if a environment was ever not specified, the minion would run just the top file for the base environment.  (default_top is set to base)  
So, we tested two configurations of env_order to see if we could enforce this behavior.  

env_order: ['base']
and
env_order: ['base', 'dev', 'env1', 'env2', 'env3', 'env4', 'env5', 'env6', 'env7, 'env8']

env_order: ['base'] resulted in the base top file being run when no environment was specified.  But, after an environment was specified on the minion, a state.apply would fail with as the minion could only see the base environment.  (state.show_top would generate the correct top file however)

env_order: ['base', 'dev', 'env1', 'env2', 'env3', 'env4', 'env5', 'env6', 'env7, 'env8'] 
This time, the correct top file was run when the environment was specified on the minion.  But, if no environment is specified, all the top files are merged.  

So, there seems to be no way to get the functionality of running only the base top file when no environment is specified and following the merge strategy same when an environment is specified.  

The above tests seem to be in conflict with the describe behavior of default_top

"When top_file_merging_strategy is set to same, and no environment is specified for a highstate (i.e. environment is not set for the minion), this config option specifies a fallback environment in which to look for a top file if an environment lacks one."

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
alt Version:
           Salt: 2017.7.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.3.1611 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.26.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core

**Comment by ccoughle** [comment]
I should also note that the topfiles for each environment only target files within that environment's file roots.  
So,, 

the base top only contains one top level env,
base:
  '*':
    - state-one

The top file for env1 would look something like

env1:
  '*':
    - state-one
    - state-two
    - state-three

No more than an single env is specified inside an env's top file

**Comment by Ch3LL** [comment]
looks like i'm able to replicate this behavior and git bisect shows this commit 82bf7fe5e1e462258963db48ed88b73f0418339e which is this pr https://github.com/saltstack/salt/pull/35822

Here is a docker container for someone to quickly replicate this:

1. `docker run -it -v ~/git/salt/:/testing/ ch3ll/issues:43293` ( where ~/git/salt/ is a local cloned git repo of salt)
2. `salt-master -d; salt-minion -d`
3. `salt '*' state.highstate`

And you will see the following:

```
1eb51ceec0f3:                            
----------                               
          ID: echo env2                  
    Function: cmd.run                    
      Result: True                       
     Comment: Command "echo env2" run    
     Started: 15:32:51.445430            
    Duration: 135.596 ms                 
     Changes:                            
              ----------                 
              pid:                       
                  376                    
              retcode:                   
                  0                      
              stderr:                    
              stdout:                    
                  env2                   
----------                               
          ID: echo env1                  
    Function: cmd.run                    
      Result: True                       
     Comment: Command "echo env1" run    
     Started: 15:32:51.581180            
    Duration: 138.542 ms                 
     Changes:                            
              ----------                 
              pid:                       
                  378                    
              retcode:                   
                  0                      
              stderr:                    
              stdout:                    
                  env1                   
----------                               
          ID: echo base                  
    Function: cmd.run                    
      Result: True                       
     Comment: Command "echo base" run    
     Started: 15:32:51.719870            
    Duration: 131.917 ms                 
     Changes:                            
              ----------                 
              pid:                       
                  380                    
              retcode:                   
                  0                      
              stderr:                    
              stdout:                    
                  base                   

Summary for 1eb51ceec0f3                 
------------                             
Succeeded: 3 (changed=3)                 
Failed:    0                             
------------                             
Total states run:     3                  
Total run time: 406.055 ms               
```

which is merging all the top files from environments env1, env2, and base

But if i run this same highstate in version 2016.3.3 it only pulls in base top file:

```
[root@31d4427db0d1 /]# salt '*' state.highstate
1eb51ceec0f3:
----------
          ID: echo base
    Function: cmd.run
      Result: True
     Comment: Command "echo base" run
     Started: 15:34:30.307752
    Duration: 143.18 ms
     Changes:   
              ----------
              pid:
                  350
              retcode:
                  0
              stderr:
              stdout:
                  base

Summary for 1eb51ceec0f3
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 143.180 ms
```

HEre are the master config settings:

```
file_roots:
  base:
    - /srv/salt/base
  env1:
    - /srv/salt/env1
    - /srv/salt/base
  env2:
    - /srv/salt/env2
    - /srv/salt/base
top_file_merging_strategy: same
default_top: base
```

Note: this also fails in 2016.11 but works in 2016.3.3 (the version i tested that is good)

We will need to get this fixed thanks.

ping @terminalmage

**Comment by terminalmage** [comment]
There seem to be some conflicting statements, here. You say that you're configuring environments in the minion's config file. But then you later mention instances in which you're not specifying an environment for the highstate. The mere act of setting the `environment` value in the minion config file sets the environment for highstates for a given minion.

**Comment by ccoughle** [comment]
@terminalmage 

We set the environment on a minion based on data in our cobbler server via external pillar calls.  I've copied the code bits below.  The instances where no environment was configured should not have happened, but because of issues with our cobbler server, the variable in the code bit below was blanked out.  This caused the errant merging behavior described above.  

{% set elienv = salt['pillar.get']('elienv') %}

eli-environment-placement-{{ elienv }}:
    file.managed:
      - name: /etc/salt/minion.d/environment.conf
      - user: root
      - group: root
      - mode: 644
      - contents:
        - "environment: {{ elienv }}"
